### PR TITLE
feat(tracing): Support for OpenTelemetry

### PR DESCRIPTION
**Pull request by aniaan**
1. Use OpenTelemetry to replace `Zipkin`
2. Support a variety of Exporter, `Zipkin`, `Jagger`, `OTLP`
3. Support trace-context and b3 context propagation formats. https://opentelemetry.io/docs/instrumentation/go/manual/#propagators-and-context
4. `tracing.tags` is replaced by `tracing.attributes`, in order to be consistent with the OpenTelemetry concept

### CODE REVIEW

File: pkg/tracing/tracing.go [hunk 1]
```diff
@@ -18,26 +18,154 @@
...
 
 type (
+	headerFormat string
 	// Spec describes Tracer.
 	Spec struct {
-		ServiceName string            `json:"serviceName" jsonschema:"required"`
-		Tags        map[string]string `json:"tags" jsonschema:"omitempty"`
-		Zipkin      *ZipkinSpec       `json:"zipkin" jsonschema:"required"`
+		ServiceName  string                `json:"serviceName" jsonschema:"required,minLength=1"`
+		Attributes   map[string]string     `json:"attributes" jsonschema:"omitempty"`
```
Comment (localvar) [inline pkg/tracing/tracing.go:46]: better keep both `Tags` and `Attributes`, and ignore `Tags` if `Attributes` is not empty, this result in better backward compatibility.

File: pkg/tracing/tracing.go [hunk 2]
```diff
@@ -48,20 +176,97 @@ type (
...
+	if spec.Jaeger == nil && spec.Zipkin == nil && spec.OTLP == nil {
+		return fmt.Errorf("exporter cannot be empty")
+	}
+
+	return nil
+}
+
+// Validate validates Spec.
+func (spec *JaegerSpec) Validate() error {
+
```
Comment (localvar) [inline pkg/tracing/tracing.go:262]: per coding style, please remove these blank lines, and the blank lines before the closing `}` of a function.

File: pkg/tracing/tracing.go [hunk 3]
```diff
@@ -83,71 +294,265 @@ func New(spec *Spec) (*Tracer, error) {
...
+	for k, v := range spec.Attributes {
+		attrs = append(attrs, attribute.String(k, v))
 	}
 
-	return &Tracer{
-		tracer: tracer,
-		closer: reporter,
-	}, nil
+	return resource.Merge(resource.Default(),
+		resource.NewWithAttributes(semconv.SchemaURL, attrs...))
```
Comment (localvar) [inline pkg/tracing/tracing.go:330]: per https://github.com/golang/go/wiki/CodeReviewComments#line-length
propose not breaking a lengthy statement into several lines.

if it is too long, try breaking it into several statements, this makes the code more readable.

there are other cases.

File: pkg/tracing/tracing.go [hunk 4]
```diff
@@ -83,71 +294,265 @@ func New(spec *Spec) (*Tracer, error) {
...
+		sampleRate = spec.Zipkin.SampleRate
+	}
+
+	if sampleRate <= 0 {
+		return sdktrace.NeverSample()
+	} else if sampleRate >= 1 {
+		return sdktrace.AlwaysSample()
+	} else {
+		return sdktrace.TraceIDRatioBased(sampleRate)
+	}
```
Comment (localvar) [inline pkg/tracing/tracing.go:346]: this makes the code more readable.

```suggestion
	if sampleRate <= 0 {
		return sdktrace.NeverSample()
	}

	if sampleRate >= 1 {
		return sdktrace.AlwaysSample()
	}

	return sdktrace.TraceIDRatioBased(sampleRate)
```

File: pkg/tracing/tracing.go [hunk 5]
```diff
@@ -83,71 +294,265 @@ func New(spec *Spec) (*Tracer, error) {
...
+			return nil, err
+		}
+	} else {
+		if exp, err := spec.Zipkin.newExporter(); err == nil {
+			exporters = []sdktrace.SpanExporter{exp}
+		} else {
+			return nil, err
+		}
+
+	}
```
Comment (localvar) [inline pkg/tracing/tracing.go:381]:
```suggestion
	} else if exp, err := spec.Zipkin.newExporter(); err == nil {
		exporters = []sdktrace.SpanExporter{exp}
	} else {
		return nil, err
	}
```

File: pkg/object/httpserver/mux.go [hunk 6]
```diff
@@ -214,7 +214,9 @@ func (mi *muxInstance) serveHTTP(stdw http.ResponseWriter, stdr *http.Request) {
 	stdr.Body = body
 
 	startAt := fasttime.Now()
-	span := mi.tracer.NewSpanWithStart(mi.superSpec.Name(), startAt)
+
+	span := mi.tracer.NewSpanWithStart(stdr.Context(), mi.superSpec.Name(), startAt)
+
```
Comment (localvar) [inline pkg/object/httpserver/mux.go:219]: @zhao-kun please help check if this change is correct or not.

**Review by localvar — APPROVED** [review]

**Review by suchen-sci — APPROVED** [review]

**Pull request merged**